### DISCUSSION TITLE
GraphQL getting filename,file content and commit date

### DISCUSSION THREAD

**Discussion by anvinraj** [community/community#24419] [Apps, API and Webhooks]
<p>I’am able to get filename, file content and commit date from github using graphql by three queries.</p>
<p>following are the querys :</p>
<pre><code># for getting filenames
query = """
        {
          repository(owner: \""""+ownerVal+"""\", name: \""""+repoVal+"""\") {
            object(expression: \""""+branchVal+":"+folderVal+"""\") {
          ... on Tree{
                entries{
                  name
                }
              }
            }
          }
        }
        """

# for getting file content
query = """
    {
      repository(owner: \""""+ownerVal+"""\", name: \""""+repoVal+"""\") {
        object(expression: \""""+branchVal+":"+folderVal+"/"+data['name']+"""\") {
             ... on Blob {
            text
          }

        }
      }
    }
    """

# for getting commit date
query = """
        {
          repository(owner: \""""+ownerVal+"""\", name: \""""+repoVal+"""\") {
            object(expression: \""""+branchVal+"""\") {
              ... on Commit {
                blame(path: \""""+folderVal+"/"+data['name']+"""\") {
                  ranges {
                    commit {
                      committedDate
                    }
                  }
                }
              }
            }
          }
        }
        """
</code></pre>
<p>How to get these three details by running one query? </p>
<p>ie, for each file in github i need to get filename, file content and commit date of the file.</p>

**Comment by lee-dohm** [comment]
<p>From what I understand, you can’t get all three details by running one query. Since <code>object</code> is a field of <code>repository</code>, you can’t request it multiple times and the expressions don’t match.</p>
<p>I hope that helps!</p>

**Comment by nickvanw** [comment, marked as answer]
<p>Hey <a class="mention" href="/anvinraj">@anvinraj</a>!</p>
<p>I believe the GraphQL API can do some of you’re looking for by using a feature called <a href="https://medium.com/graphql-mastery/graphql-quick-tip-aliases-567303a9ddc5" rel="nofollow noopener">aliasing</a>, which allows you to give connections a unique name, and execute multiple of the same connections in a single query. Take your example run against one of my Repositories:</p>
<pre><code>query {
  repository(owner: "nickvanw", name: "ircx") {
    commit: object(expression: "49cff0c1f801176c4cdf774252a5803f25cdb815") {
      ... on Commit {
        oid
      }
    }

    content:object(expression: "master:README.md") {
      ... on Blob {
        text
      }
    }

    filename: object(expression: "master:example/") {
      ... on Tree {
        entries {
          name
        }
      }
    }
  }
}
</code></pre>
<p>This returns all three of them, in one query:</p>
<pre><code>{
  "data": {
    "repository": {
      "commit": {
        "oid": "49cff0c1f801176c4cdf774252a5803f25cdb815"
      },
      "content": {
        "text": "ircx\n====\n[![Build Status](https://travis-ci.org/nickvanw/ircx.svg?branch=master)](https://travis-ci.org/nickvanw/ircx)\n\nircx is a very basic IRC bot written on top of the wonderfully small [sorcix/irc](https://github.com/sorcix/irc) library. It's designed to be a small building block, a small example of one way to use the library.\n\nUsing it is very simple, see [the example](example/main.go).\n\nThis example doesn't do anything other than connect to specified channels and idle, but it's trivial to add additional handlers for any IRC event you want.\n\nContext can be passed around by creating custom Handlers and Senders and using them, versus the default sender created, and an empty handler struct.\n"
      },
      "filename": {
        "entries": [
          {
            "name": "main.go"
          }
        ]
      }
    }
  }
}
</code></pre>
<p>The only limitation in this is that you cannot use the output of one query into an input field in the same query. However, I do think that there are some optimizations to your query that can be done. For example, I believe you can go directly from the Tree to the object contents, by doing something like this:</p>
<pre><code>query {
  repository(owner: "...", name: "...") {
    folder: object(expression: "master:path/to/folder/") {
      ... on Tree {
        entries {
          oid
          object {
            ... on Blob {
              text
            }
          }
          name
        }
      }
    }
  }
}
</code></pre>
<p>Let me know if this makes sense, and if anything else looks amiss! I would also recommend using <a href="https://medium.com/the-graphqlhub/graphql-tour-variables-58c6abd10f56" rel="nofollow noopener">GraphQL Variables</a>, which would allow you to paramaterize some of what you’re doing, instead of string concatenation.</p>
<p>Let me know if you have any more questions!</p>

**Reply by tonglil** [reply]
I'm wondering if anyone has ever attempted this with files that are symlinks?
I can't find any documentation on how to follow the symlinks, which show up as 
```
            "object": {
              "text": "../../link-to-file"
            },
```
where as I'd like to get the contents of the file at the symlink.

The REST API resolves the symlink, but the graphql one does not.

**Comment by Biboswan** [comment]
<pre><code>{
  repository(owner:"imstack", name:"taskcluster-repoifier" ) {
    object(expression: "master:README.md") {
      ...on Blob {
      text
      byteSize
    }
  }
}
</code></pre>
<p>I tried some queries but getting response as below? Any clue pls?</p>
<pre><code>{
  "errors": [
    {
      "message": "Unexpected end of document"
    }
  ]
}
</code></pre>

**Comment by appernetic** [comment]
<p>Yes! You have a missing } at the end.</p>
<p>{<br>
repository(owner:“appernetic”, name:“hugo-bootstrap-premium-dev” ) {<br>
object(expression: “master:README.md”) {<br>
…on Blob {<br>
text<br>
byteSize<br>
               }<br>
      }<br>
}<br>
}</p>

**Comment by bialikover** [comment]
<p>Is there any way to get only the files changed in that specific commit?</p>

**Reply by thompsch** [reply]
Did you ever get this figured out?

**Comment by pcm1289** [comment]
<blockquote>
<p><a class="mention" href="/bialikover">@bialikover</a> wrote:</p>
</blockquote>
<blockquote>
<p>Is there any way to get only the files changed in that specific commit?</p>
</blockquote>
<p>Just kidding, this query gets all files in the repo…</p>
<pre><code>query {
  repository(owner: "nickvanw", name: "ircx") {
    commit: object(expression: "49cff0c1f801176c4cdf774252a5803f25cdb815") {
      ... on Commit {
        tree {
          entries { name }
        }
      }
    }
  }
}
</code></pre>
